File: Leetcode 3901-4000/3956/3956.java
// Leetcode 3956: Maximum Sum of M Non-Overlapping Subarrays I
// https://leetcode.com/problems/maximum-sum-of-m-non-overlapping-subarrays-i/
// Solved on 25th of June, 2026
class Solution {
    /**
     * Computes the maximum sum of m non-overlapping subarrays of length between l and r.
     * @param nums The input array.
     * @param m The number of subarrays.
     * @param l The minimum length of a subarray.
     * @param r The maximum length of a subarray.
     * @return The maximum sum of m non-overlapping subarrays.
     */
    public long maximumSum(int[] nums, int m, int l, int r) {
        int n = nums.length;
        long[] prefixSum = new long[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }
        long minVal = -10000000000000000L;
        long[] prevDp = new long[n + 1];
        long finalMaxSum = minVal;
        for (int c = 1; c <= m; c++) {
            long[] currDp = new long[n + 1];
            for (int i = 0; i <= n; i++) {
                currDp[i] = minVal;
            }
            int[] deque = new int[n + 1];
            int head = 0;
            int tail = 0;
            for (int i = 1; i <= n; i++) {
                int right = i - l;
                int left = i - r;
                if (right >= 0 && prevDp[right] != minVal) {
                    long val = prevDp[right] - prefixSum[right];
                    while (head < tail && prevDp[deque[tail - 1]] - prefixSum[deque[tail - 1]] <= val) {
                        tail--;
                    }
                    deque[tail++] = right;
                }
                while (head < tail && deque[head] < left) {
                    head++;
                }
                currDp[i] = currDp[i - 1];
                if (head < tail) {
                    int bestJ = deque[head];
                    long sum = prefixSum[i] + prevDp[bestJ] - prefixSum[bestJ];
                    if (sum > currDp[i]) {
                        currDp[i] = sum;
                    }
                }
            }
            if (currDp[n] > finalMaxSum) {
                finalMaxSum = currDp[n];
            }
            prevDp = currDp;
        }
        return finalMaxSum;
    }
}

File: Leetcode 3901-4000/3956/3956.py
# Leetcode 3956: Maximum Sum of M Non-Overlapping Subarrays I
# https://leetcode.com/problems/maximum-sum-of-m-non-overlapping-subarrays-i/
# Solved on 25th of June, 2026
import collections


class Solution:
    def maximumSum(self, nums: list[int], m: int, l: int, r: int) -> int:
        """
        Computes the maximum sum of m non-overlapping subarrays of length between l and r.
        
        :param nums: The input array.
        :param m: The number of subarrays.
        :param l: The minimum length of a subarray.
        :param r: The maximum length of a subarray.
        :return: The maximum sum of m non-overlapping subarrays.
        """
        n = len(nums)
        prefixSum = [0] * (n + 1)
        for i in range(n):
            prefixSum[i + 1] = prefixSum[i] + nums[i]
        
        prevDp = [0] * (n + 1)
        maxAns = float('-inf')

        for c in range(1, m + 1):
            currDp = [float('-inf')] * (n + 1)
            dq = collections.deque()

            for i in range(l, n + 1):
                currJ = i - l

                if prevDp[currJ] != float('-inf'):
                    val = prevDp[currJ] - prefixSum[currJ]
                    while dq and dq[-1][1] <= val:
                        dq.pop()
                    dq.append((currJ, val))

                while dq and dq[0][0] < i - r:
                    dq.popleft()

                currDp[i] = currDp[i - 1]

                if dq:
                    currDp[i] = max(currDp[i], prefixSum[i] + dq[0][1])

            maxAns = max(maxAns, currDp[n])
            prevDp = currDp

        return int(maxAns)